pkg/insightstore/chunk.go

```diff
@@ -0,0 +1,280 @@
+	case model.InsightStep_WEEKLY:
+		return toDataPoints(c.DataPoints.Weekly)
+	case model.InsightStep_DAILY:
+		return toDataPoints(c.DataPoints.Daily)
+	}
+	return []DataPoint{}, fmt.Errorf("invalid step: %v", step)
+}
+
+func (c *ChangeFailureRateChunk) DataCount(step model.InsightStep) int {
+	switch step {
+	case model.InsightStep_YEARLY:
+		return len(c.DataPoints.Yearly)
+	case model.InsightStep_MONTHLY:
+		return len(c.DataPoints.Monthly)
+	case model.InsightStep_WEEKLY:
+		return len(c.DataPoints.Weekly)
+	case model.InsightStep_DAILY:
+		return len(c.DataPoints.Daily)
+	}
+	return 0
+}
+
+func (c *ChangeFailureRate) GetTimestamp() int64 {
+	return c.Timestamp
+}
+
+func (c *ChangeFailureRate) Value() float32 {
+	return c.Rate
+}
+
+type Chunk interface {
+	// GetFilePath gets filepath
+	GetFilePath() string
+	// SetFilePath sets filepath
+	SetFilePath(path string)
+	// GetAccumulatedTo gets AccumulatedTo
+	GetAccumulatedTo() int64
+	// SetAccumulatedTo sets AccumulatedTo
+	SetAccumulatedTo(a int64)
+	// GetDataPoint gets list of data points of specify step
+	GetDataPoint(step model.InsightStep) ([]DataPoint, error)
+	// DataCount returns count of data in specify step
+	DataCount(step model.InsightStep) int
+}
+
+// convert types to Chunk.
+func toChunk(i interface{}) (Chunk, error) {
+	switch p := i.(type) {
+	case *DeployFrequencyChunk:
+		return p, nil
+	case *ChangeFailureRateChunk:
+		return p, nil
+	default:
+		return nil, fmt.Errorf("cannot convert to Chunk: %v", p)
+	}
+}
+
+type DataPoint interface {
+	// Value gets data for model.InsightDataPoint
+	Value() float32
+	// Timestamp gets timestamp
+	GetTimestamp() int64
+}
+
+// convert types to list of DataPoint.
+func toDataPoints(i interface{}) ([]DataPoint, error) {
+	switch dps := i.(type) {
+	case []*DeployFrequency:
+		dataPoints := make([]DataPoint, len(dps))
+		for j, dp := range dps {
+			dataPoints[j] = dp
+		}
+		return dataPoints, nil
+	case []*ChangeFailureRate:
+		dataPoints := make([]DataPoint, len(dps))
+		for j, dp := range dps {
+			dataPoints[j] = dp
+		}
+		return dataPoints, nil
+	default:
+		return nil, fmt.Errorf("cannot convert to DataPoints: %v", dps)
+	}
+}
+
+type Chunks []Chunk
+
+func (cs Chunks) ExtractDataPoints(step model.InsightStep, from time.Time, count int) ([]*model.InsightDataPoint, error) {
+	var idps []*model.InsightDataPoint
+	for _, c := range cs {
+		idp, err := chunkToDataPoints(c, from, count, step)
+		if err != nil {
+			return nil, err
+		}
+
+		idps = append(idps, idp...)
+
+		count = count - c.DataCount(step)
+
+		nextMonth := time.Date(from.Year(), from.Month()+1, 1, 0, 0, 0, 0, time.UTC)
+		from = normalizeTime(nextMonth, step)
+		if step == model.InsightStep_WEEKLY && from.Month() != nextMonth.Month() {
+			from = from.AddDate(0, 0, 7)
+		}
+	}
+
+	return idps, nil
+}
+
+func chunkToDataPoints(chunk Chunk, from time.Time, count int, step model.InsightStep) ([]*model.InsightDataPoint, error) {
```

Comment:
nit: `extractDataPoints(chunk Chunk, step model.InsightStep, from, to time.Time)`